File: src/main/java/com/codebutler/farebot/transit/ovc/OVChipTransaction.java
/*
 * OVChipTransaction.java
 *
 * This file is part of FareBot.
 * Learn more at: https://codebutler.github.io/farebot/
 *
 * Copyright (C) 2012 Wilbert Duijvenvoorde <w.a.n.duijvenvoorde@gmail.com>
 * Copyright (C) 2012, 2014-2015 Eric Butler <eric@codebutler.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.codebutler.farebot.transit.ovc;

import android.os.Parcelable;
import android.support.annotation.NonNull;

import com.codebutler.farebot.util.Utils;
import com.google.auto.value.AutoValue;

import java.util.Comparator;

@AutoValue
abstract class OVChipTransaction implements Parcelable {

    static final Comparator<OVChipTransaction> ID_ORDER = new Comparator<OVChipTransaction>() {
        @Override
        public int compare(OVChipTransaction t1, OVChipTransaction t2) {
            return (t1.getId() < t2.getId() ? -1 : (t1.getId() == t2.getId() ? 0 : 1));
        }
    };

    @NonNull
    static OVChipTransaction create(int transactionSlot, byte[] data) {
        if (data == null) {
            data = new byte[32];
        }

        int valid = 1;
        int date = 0;
        int time = 0;
        int unknownConstant = 0;
        int transfer = -3; // Default: No-data
        int company = 0;
        int id = 0;
        int station = 0;
        int machineId = 0;
        int vehicleId = 0;
        int productId = 0;
        int unknownConstant2 = 0;
        int amount = 0;
        int subscriptionId = -1; // Default: No valid subscriptionId
        String errorMessage = "";

        if (data[0] == (byte) 0x00
                && data[1] == (byte) 0x00
                && data[2] == (byte) 0x00
                && (data[3] & (byte) 0xF0) == (byte) 0x00) {
            valid = 0;
        }
        if ((data[3] & (byte) 0x10) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[3] & (byte) 0x80) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[2] & (byte) 0x02) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[2] & (byte) 0x08) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[2] & (byte) 0x20) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[2] & (byte) 0x80) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[1] & (byte) 0x01) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[1] & (byte) 0x02) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[1] & (byte) 0x08) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[1] & (byte) 0x20) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[1] & (byte) 0x40) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[1] & (byte) 0x80) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[0] & (byte) 0x02) != (byte) 0x00) {
            valid = 0;
        }
        if ((data[0] & (byte) 0x04) != (byte) 0x00) {
            valid = 0;
        }

        if (valid == 0) {
            errorMessage = "No transaction";
        } else {
            int iBitOffset = 53; // Ident, Date, Time

            date = (((char) data[3] & (char) 0x0F) << 10) | (((char) data[4] & (char) 0xFF) << 2)
                    | (((char) data[5] >> 6) & (char) 0x03);
            time = (((char) data[5] & (char) 0x3F) << 5) | (((char) data[6] >> 3) & (char) 0x1F);

            if ((data[3] & (byte) 0x20) != (byte) 0x00) {
                unknownConstant = Utils.getBitsFromBuffer(data, iBitOffset, 24);
                iBitOffset += 24;
            }

            if ((data[3] & (byte) 0x40) != (byte) 0x00) {
                transfer = Utils.getBitsFromBuffer(data, iBitOffset, 7);
                iBitOffset += 7;
            }

            if ((data[2] & (byte) 0x01) != (byte) 0x00) {
                company = Utils.getBitsFromBuffer(data, iBitOffset, 16);
                iBitOffset += 16;
            }

            if ((data[2] & (byte) 0x04) != (byte) 0x00) {
                id = Utils.getBitsFromBuffer(data, iBitOffset, 24);
                iBitOffset += 24;
            }

            if ((data[2] & (byte) 0x10) != (byte) 0x00) {
                station = Utils.getBitsFromBuffer(data, iBitOffset, 16);
                iBitOffset += 16;
            }

            if ((data[2] & (byte) 0x40) != (byte) 0x00) {
                machineId = Utils.getBitsFromBuffer(data, iBitOffset, 24);
                iBitOffset += 24;
            }

            if ((data[1] & (byte) 0x04) != (byte) 0x00) {
                vehicleId = Utils.getBitsFromBuffer(data, iBitOffset, 16);
                iBitOffset += 16;
            }

            if ((data[1] & (byte) 0x10) != (byte) 0x00) {
                productId = Utils.getBitsFromBuffer(data, iBitOffset, 5);
                iBitOffset += 5;
            }

            if ((data[0] & (byte) 0x01) != (byte) 0x00) {
                unknownConstant2 = Utils.getBitsFromBuffer(data, iBitOffset, 16);
                iBitOffset += 16;
            }

            if ((data[0] & (byte) 0x08) != (byte) 0x00) {
                amount = Utils.getBitsFromBuffer(data, iBitOffset, 16);
                iBitOffset += 16;
            }

            if ((data[1] & (byte) 0x10) == (byte) 0x00) {
                subscriptionId = Utils.getBitsFromBuffer(data, iBitOffset, 13);
            }
        }

        return new AutoValue_OVChipTransaction.Builder()
                .date(date)
                .time(time)
                .transfer(transfer)
                .company(company)
                .id(id)
                .station(station)
                .machineId(machineId)
                .vehicleId(vehicleId)
                .productId(productId)
                .amount(amount)
                .subscriptionId(subscriptionId)
                .valid(valid)
                .unknownConstant(unknownConstant)
                .unknownConstant2(unknownConstant2)
                .transactionSlot(transactionSlot)
                .errorMessage(errorMessage)
                .build();
    }

    public abstract int getTransactionSlot();

    public abstract int getDate();

    public abstract int getTime();

    public abstract int getTransfer();

    public abstract int getCompany();

    public abstract int getId();

    public abstract int getStation();

    public abstract int getMachineId();

    public abstract int getVehicleId();

    public abstract int getProductId();

    public abstract int getAmount();

    public abstract int getSubscriptionId();

    public abstract int getValid();

    public abstract int getUnknownConstant();

    public abstract int getUnknownConstant2();

    public abstract String getErrorMessage();

    public boolean isSameTrip(OVChipTransaction nextTransaction) {
        /*
         * Information about checking in and out:
         * http://www.chipinfo.nl/inchecken/
         */

        if (getCompany() == nextTransaction.getCompany() && getTransfer() == OVChipTransitData.PROCESS_CHECKIN
                && nextTransaction.getTransfer() == OVChipTransitData.PROCESS_CHECKOUT) {
            if (getDate() == nextTransaction.getDate()) {
                return true;
            } else if (getDate() == nextTransaction.getDate() - 1) {
                // All NS trips get reset at 4 AM (except if it's a night train, but that's out of our scope).
                if (getCompany() == OVChipTransitData.AGENCY_NS && nextTransaction.getTime() < 240) {
                    return true;
                }

                /*
                 * Some companies expect a checkout at the maximum of 15 minutes after the estimated arrival at the
                 * endstation of the line.
                 * But it's hard to determine the length of every single trip there is, so for now let's just assume a
                 * checkout at the next day is still from the same trip. Better solutions are always welcome ;)
                 */
                if (getCompany() != OVChipTransitData.AGENCY_NS) {
                    return true;
                }
            }
        }

        return false;
    }

    @AutoValue.Builder
    abstract static class Builder {
        abstract Builder transactionSlot(int transactionSlot);

        abstract Builder date(int date);

        abstract Builder time(int time);

        abstract Builder transfer(int transfer);

        abstract Builder company(int company);

        abstract Builder id(int id);

        abstract Builder station(int station);

        abstract Builder machineId(int machineId);

        abstract Builder vehicleId(int vehicleId);

        abstract Builder productId(int productId);

        abstract Builder amount(int amount);

        abstract Builder subscriptionId(int subscriptionId);

        abstract Builder valid(int valid);

        abstract Builder unknownConstant(int unknownConstant);

        abstract Builder unknownConstant2(int unknownConstant2);

        abstract Builder errorMessage(String errorMessage);

        abstract OVChipTransaction build();
    }
}

File: farebot-transit-ovc/src/commonMain/kotlin/com/codebutler/farebot/transit/ovc/OVChipTransaction.kt
/*
 * OVChipTransaction.kt
 *
 * Copyright 2012 Wilbert Duijvenvoorde <w.a.n.duijvenvoorde@gmail.com>
 * Copyright 2012 Eric Butler <eric@codebutler.com>
 * Copyright 2025 Eric Butler <eric@codebutler.com>
 *
 * Ported from Metrodroid (https://github.com/metrodroid/metrodroid)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.codebutler.farebot.transit.ovc

import com.codebutler.farebot.base.util.isAllZero
import com.codebutler.farebot.transit.Transaction
import com.codebutler.farebot.transit.Trip
import com.codebutler.farebot.transit.en1545.En1545Bitmap
import com.codebutler.farebot.transit.en1545.En1545Container
import com.codebutler.farebot.transit.en1545.En1545FixedHex
import com.codebutler.farebot.transit.en1545.En1545FixedInteger
import com.codebutler.farebot.transit.en1545.En1545Lookup
import com.codebutler.farebot.transit.en1545.En1545Parsed
import com.codebutler.farebot.transit.en1545.En1545Parser
import com.codebutler.farebot.transit.en1545.En1545Transaction
import com.codebutler.farebot.transit.en1545.getBitsFromBuffer

class OVChipTransaction(override val parsed: En1545Parsed) : En1545Transaction() {
    private val date: Int
        get() = parsed.getIntOrZero(En1545FixedInteger.dateName(EVENT))

    private val time: Int
        get() = parsed.getIntOrZero(En1545FixedInteger.timeLocalName(EVENT))

    private val transfer: Int
        get() = parsed.getIntOrZero(TRANSACTION_TYPE)

    private val company: Int
        get() = parsed.getIntOrZero(EVENT_SERVICE_PROVIDER)

    val id: Int
        get() = parsed.getIntOrZero(EVENT_SERIAL_NUMBER)

    override val lookup: En1545Lookup get() = OvcLookup

    override val isTapOn get() = transfer == PROCESS_CHECKIN

    override val isTapOff get() = transfer == PROCESS_CHECKOUT

    override fun isSameTrip(other: Transaction): Boolean {
        if (other !is OVChipTransaction)
            return false
        /*
         * Information about checking in and out:
         * http://www.chipinfo.nl/inchecken/
         */

        if (company != other.company)
            return false
        if (date == other.date) {
            return true
        }
        if (date != other.date - 1)
            return false

        // All NS trips get reset at 4 AM (except if it's a night train, but that's out of our scope).
        if (company == AGENCY_NS) {
            return other.time < 240
        }

        /*
         * Some companies expect a checkout at the maximum of 15 minutes after the estimated arrival at the
         * endstation of the line.
         * But it's hard to determine the length of every single trip there is, so for now let's just assume a
         * checkout at the next day is still from the same trip. Better solutions are always welcome ;)
         */
        return true
    }

    override val mode get(): Trip.Mode {
        val startStationId = stationId ?: 0

        // FIXME: Clean this up
        //mIsBusOrTram = (company == AGENCY_GVB || company == AGENCY_HTM || company == AGENCY_RET && (!mIsMetro));
        //mIsBusOrTrain = company == AGENCY_VEOLIA || company == AGENCY_SYNTUS;

        when (transfer) {
            PROCESS_BANNED -> return Trip.Mode.BANNED
            PROCESS_CREDIT -> return Trip.Mode.TICKET_MACHINE
            // Not 100% sure about what NODATA is, but looks alright so far
            PROCESS_PURCHASE, PROCESS_NODATA -> return Trip.Mode.TICKET_MACHINE
        }

        return when (company) {
            AGENCY_NS -> Trip.Mode.TRAIN
            AGENCY_TLS, AGENCY_DUO, AGENCY_STORE -> Trip.Mode.OTHER
            // TODO: Needs verification!
            AGENCY_GVB -> if (startStationId < 3000) Trip.Mode.METRO else Trip.Mode.BUS
            // TODO: Needs verification!
            AGENCY_RET -> if (startStationId < 3000) Trip.Mode.METRO else Trip.Mode.BUS
            AGENCY_ARRIVA -> when (startStationId) {
                in 0..800 -> Trip.Mode.TRAIN
                // TODO: Needs verification!
                in 4601..4699 -> Trip.Mode.FERRY
                else -> Trip.Mode.BUS
            }
            // Everything else will be a bus, although this is not correct.
            // The only way to determine them would be to collect every single 'ovcid' out there :(
            else -> Trip.Mode.BUS
        }
    }

    companion object {
        private const val PROCESS_PURCHASE = 0x00
        private const val PROCESS_CHECKIN = 0x01
        private const val PROCESS_CHECKOUT = 0x02
        private const val PROCESS_TRANSFER = 0x06
        private const val PROCESS_BANNED = 0x07
        private const val PROCESS_CREDIT = -0x02
        private const val PROCESS_NODATA = -0x03

        private const val AGENCY_TLS = 0x00
        private const val AGENCY_GVB = 0x02
        private const val AGENCY_NS = 0x04
        private const val AGENCY_RET = 0x05
        private const val AGENCY_ARRIVA = 0x08
        private const val AGENCY_DUO = 0x0C    // Could also be 2C though... ( http://www.ov-chipkaart.me/forum/viewtopic.php?f=10&t=299 )
        private const val AGENCY_STORE = 0x19
        private const val AGENCY_GVB_FLEX = 0x2711

        const val TRANSACTION_TYPE = "TransactionType"

        private fun neverSeen(i: Int) = "NeverSeen$i"

        private fun neverSeenField(i: Int) = En1545FixedInteger(neverSeen(i), 8)

        fun tripFields(reversed: Boolean = false) = En1545Bitmap.infixBitmap(
                En1545Container(
                        En1545FixedInteger.date(EVENT),
                        En1545FixedInteger.timeLocal(EVENT)
                ),
                neverSeenField(1),
                En1545FixedInteger(EVENT_UNKNOWN_A, 24),
                En1545FixedInteger(TRANSACTION_TYPE, 7),
                neverSeenField(4),
                En1545FixedInteger(EVENT_SERVICE_PROVIDER, 16),
                neverSeenField(6),
                En1545FixedInteger(EVENT_SERIAL_NUMBER, 24),
                neverSeenField(8),
                En1545FixedInteger(EVENT_LOCATION_ID, 16),
                neverSeenField(10),
                En1545FixedInteger(EVENT_DEVICE_ID, 24),
                neverSeenField(12),
                neverSeenField(13),
                neverSeenField(14),
                En1545FixedInteger(EVENT_VEHICLE_ID, 16),
                neverSeenField(16),
                En1545FixedInteger(EVENT_CONTRACT_POINTER, 5),
                neverSeenField(18),
                neverSeenField(19),
                neverSeenField(20),
                En1545FixedInteger("TripDurationMinutes", 16),
                neverSeenField(22),
                neverSeenField(23),
                En1545FixedInteger(EVENT_PRICE_AMOUNT, 16),
                En1545FixedInteger("EventSubscriptionID", 13),
                // Could be from 8 to 10
                En1545FixedInteger(EVENT_UNKNOWN_C, 10),
                neverSeenField(27),
                En1545FixedInteger("EventExtra", 0),
                reversed = reversed
        )

        private val OVC_UL_TRIP_FIELDS = En1545Container(
                En1545FixedInteger("A", 8),
                En1545FixedInteger(EVENT_SERIAL_NUMBER, 12),
                En1545FixedInteger(EVENT_SERVICE_PROVIDER, 12),
                En1545FixedInteger(TRANSACTION_TYPE, 3),
                En1545FixedInteger.date(EVENT),
                En1545FixedInteger.timeLocal(EVENT),
                En1545FixedInteger("balseqno", 4),
                En1545FixedHex("D", 64)
        )

        fun parseClassic(data: ByteArray): OVChipTransaction? {
            if (data.getBitsFromBuffer(0, 28) == 0)
                return null
            val parsed = En1545Parser.parse(data, tripFields())
            // 27 is not critical, ignore if ever
            for (i in 1..23)
                if (parsed.contains(neverSeen(i)))
                    return null
            return OVChipTransaction(parsed)
        }

        fun parseUltralight(data: ByteArray): OVChipTransaction? {
            if (data.isAllZero())
                return null
            return OVChipTransaction(En1545Parser.parse(data, OVC_UL_TRIP_FIELDS))
        }
    }
}
